### DISCUSSION TITLE
Automatic header parsing

### DISCUSSION THREAD

**Discussion by vigji** [bonsai-rx/bonsai#1616] [Q&A]
I am often in the situation of logging in csv files all the values assumed by some behavior subject to which I subscribe with a CsvWriter node, usually timestamped. I end up every time adding an ExpressionTransform node to define the headers of the new `.csv` file, as suggested many times in the Bonsai Q&A.

I was wondering if there is a way of implementing the same behavior using directly the Subject name as the header, ideally both in the scenario where I am logging a single variable or where I am grouping multiple values in a tuple before saving it.

If there is no such option in "vanilla Bonsai", would there be a way of getting there via a custom scripted node?

Thank you in advance for the answer, and the amazing tool

**Comment by bruno-f-cruz** [comment]
There is no easy way of achieving this currently. Could you also provide an example why having a dynamic header would be useful? I usually look at headers as something that provide a somewhat stable and predictable structure, so I am wondering if there would not be a better way to go around this perhaps? Maybe changing the file name?

**Reply by vigji** [reply]
Dear Bruno, thank you very much for your answer.

Probably I was not explaining well (or maybe my considerations are a bit off - I am not a great expert), but I do not need a dynamic header as in, something that changes mid-flow during a workflow execution. Instead, I would like to understand if there are ways to minimize inpoints of variable names in the experiment, to reduce bugs.

Minimal example, I am using something like this flow to save values that are streamed by an external application via ZMQ to two variables that I use elsewhere in the experiment. I also want to log their values to a csv, so I do:
![example_flow](https://github.com/bonsai-rx/bonsai/assets/10400212/2a388fd8-5321-467f-8dd2-fc0af783918c)

Where in the ExpressionTransform I use:

```C#
new(Value.Item1 as X,
    Value.Item2 as Y,
    Timestamp as Timestamp)
```

Now, maybe this is stupid, but I am annoyed by the fact of having to duplicate the variable names (X, Y) in the expression transform. What makes me believe it could be less stupid is because:
 - sometimes I change the variable names and I forgot to update the headers names, leading to mistakes
 - not defining those values programmatically is what is currently preventing me from trying and make a little extension workflow out of the saving pattern that I use a lot
 - Because of 2, copypasing the pattern around it has happened to me multiple times to forget to update the ExpressionTransform.

I could use the filename instead of the header and log each variable separately, but same issue there: I would like to have filenames automatically matching variable names and I would not know how to do that.

I hope that it is now clearer what I am aiming for - if it makes sense at all! And thanks again for your help!

**Reply by bruno-f-cruz** [reply]
I don't see an easy way to do this using the current bonsai semantics. The closest thing I can think of would probably be to define your own serialization function to json.


https://learn.microsoft.com/en-us/dotnet/standard/serialization/system-text-json/how-to?pivots=dotnet-8-0

I guess you could also just use a Dictionary<paramName,paramValue> and dump that to a text file. Then use the first value to write the first line, using the keys as headers. This would always require you to write your own csharp code.

Maybe @glopesdev has an easier alternative?

Let me know if you end up finding a solution, I am curious :)

**Comment by vigji** [comment]
I surely will, thanks for your help! I'll leave this open while I work on it in case there's other inputs that might be coming.

**Comment by petersaj** [comment]
I just stumbled on this thread, I had a similar issue and solved it with a similarly silly set of nodes (https://github.com/orgs/bonsai-rx/discussions/1271). 

I'd be very interested if there were any updates on this: using a node/expression to pull out the name of a subject would be extremely useful for saving data.
